cpp/tests/join/join_tests.cu

```diff
@@ -0,0 +1,456 @@
+}
+
+TEST_F(JoinTest, InnerJoin)
+{
+  column_wrapper <int32_t> col0_0{{3, 1, 2, 0, 3}};
+  column_wrapper <int32_t> col0_1{{0, 1, 2, 4, 1}};
+
+  column_wrapper <int32_t> col1_0{{2, 2, 0, 4, 3}};
+  column_wrapper <int32_t> col1_1{{1, 0, 1, 2, 1}};
+
+  CVector cols0, cols1;
+  cols0.push_back(col0_0.release());
+  cols0.push_back(col0_1.release());
+  cols1.push_back(col1_0.release());
+  cols1.push_back(col1_1.release());
+
+  Table t0(std::move(cols0));
+  Table t1(std::move(cols1));
+
+  auto join_table = cudf::inner_join(t0, t1, {0, 1}, {0, 1}, {{0, 0}, {1,1}});
+  cudf::test::print(join_table->get_column(0)); std::cout<<"\n";
+  cudf::test::print(join_table->get_column(1)); std::cout<<"\n";
+}
+
+TEST_F(JoinTest, LeftSemiJoin) {
+  std::vector<const char*> a_strings{ "quick", "accénted", "turtlé", "composéd", "result", "", "words" };
+  std::vector<const char*> b_strings { "quick", "words", "result" };
+  std::vector<const char*> e_strings { "quick", "composéd", "result", "" };
+
+  column_wrapper<int32_t> a_0 {  10,  20,  20,  20,  20,  20,  50 };
+  column_wrapper<float>   a_1 { 5.0,  .5,  .5,  .7,  .7,  .7,  .7 };
+  column_wrapper<int8_t>  a_2 {  90,  77,  78,  61,  62,  63,  41 };
+
+  cudf::test::strings_column_wrapper a_3(a_strings.begin(),
+                                         a_strings.end(),
+                                         thrust::make_transform_iterator( a_strings.begin(), [] (auto str) { return str!=nullptr; }));
+
+  column_wrapper<int32_t> b_0 {  10, 20, 20 };
+  column_wrapper<float>   b_1 { 5.0, .7, .7 };
+  column_wrapper<int8_t>  b_2 {  90, 75, 62 };
+
+  cudf::test::strings_column_wrapper b_3(b_strings.begin(),
+                                         b_strings.end(),
+                                         thrust::make_transform_iterator( b_strings.begin(), [] (auto str) { return str!=nullptr; }));
+
+
+  column_wrapper<int32_t> expect_0 {  10,  20,  20,  20 };
+  column_wrapper<float>   expect_1 { 5.0,  .7,  .7,  .7 };
+  column_wrapper<int8_t>  expect_2 {  90,  61,  62,  63 };
+
+  cudf::test::strings_column_wrapper expect_3(e_strings.begin(),
+                                              e_strings.end(),
+                                              thrust::make_transform_iterator( e_strings.begin(), [] (auto str) { return str!=nullptr; }));
+
+
+  std::vector<std::unique_ptr<cudf::column>> column_a;
+  column_a.push_back(a_0.release());
+  column_a.push_back(a_1.release());
+  column_a.push_back(a_2.release());
+  column_a.push_back(a_3.release());
+
+  std::vector<std::unique_ptr<cudf::column>> column_b;
+  column_b.push_back(b_0.release());
+  column_b.push_back(b_1.release());
+  column_b.push_back(b_2.release());
+  column_b.push_back(b_3.release());
+
+  cudf::experimental::table table_a(std::move(column_a));
+  cudf::experimental::table table_b(std::move(column_b));
+
+  auto join_table = cudf::left_semi_join(table_a, table_b,
```

Comment:
I can't tell, do you test the error cases for semi/anti joins? What about the empty input cases? We should exercise these pathways.

Reply:
Not yet.  I can add those.